librad/src/git/local/transport.rs

```diff
@@ -0,0 +1,426 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    io::{self, Cursor, Read, Write},
+    path::PathBuf,
+    process::{Child, ChildStdin, ChildStdout, Command, ExitStatus, Stdio},
+    sync::{Arc, Mutex, Once, RwLock},
+    thread,
+};
+
+use git2::transport::{Service, SmartSubtransport, SmartSubtransportStream, Transport};
+use thiserror::Error;
+
+use crate::{
+    git::{
+        ext::{into_git_err, RECEIVE_PACK_HEADER, UPLOAD_PACK_HEADER},
+        local::{self, url::LocalUrl},
+        storage::{self, Storage, WithSigner},
+    },
+    keys::SecretKey,
+    paths::Paths,
+    uri::RadUrn,
+};
+
+lazy_static! {
+    static ref SETTINGS: Arc<RwLock<Option<Settings>>> = Arc::new(RwLock::new(None));
+}
+
+#[derive(Clone)]
+pub struct Settings {
+    pub paths: Paths,
+    pub signer: SecretKey,
+}
+
+pub fn register(settings: Settings) {
+    static INIT: Once = Once::new();
+
+    LocalTransportFactory::new().configure(settings);
+    unsafe {
+        INIT.call_once(move || {
+            git2::transport::register(local::URL_SCHEME, move |remote| {
+                Transport::smart(&remote, true, LocalTransportFactory::new())
+            })
+            .unwrap()
+        });
+    }
+}
+
+#[derive(Clone)]
+struct LocalTransportFactory {
+    settings: Arc<RwLock<Option<Settings>>>,
+}
+
+impl LocalTransportFactory {
+    fn new() -> Self {
+        Self {
+            settings: SETTINGS.clone(),
+        }
+    }
+
+    fn configure(&self, settings: Settings) {
+        *self.settings.write().unwrap() = Some(settings)
+    }
+}
+
+impl SmartSubtransport for LocalTransportFactory {
+    fn action(
+        &self,
+        url: &str,
+        service: Service,
+    ) -> Result<Box<dyn SmartSubtransportStream>, git2::Error> {
+        let settings = self.settings.read().unwrap();
+        match *settings {
+            None => Err(git2::Error::from_str("local transport unconfigured")),
+            Some(ref settings) => {
+                let url = url.parse::<LocalUrl>().map_err(into_git_err)?;
+
+                let mut transport = LocalTransport::new(settings.clone()).map_err(into_git_err)?;
+                let stream = transport
+                    .stream(url, service, Localio::piped())
+                    .map_err(into_git_err)?;
+
+                Ok(Box::new(stream))
+            },
+        }
+    }
+
+    fn close(&self) -> Result<(), git2::Error> {
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Mode {
+    Stateless,
+    Stateful,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("No such URN: {0}")]
+    NoSuchUrn(RadUrn),
+
+    #[error(transparent)]
+    Storage(#[from] storage::Error),
+
+    #[error("Child exited unsuccessfully")]
```

Comment:
I considered it, but you can't call a method from the `thiserror` macros. I don't think it's worth moving out the error code, because whether it's `42` or `255` is most likely not going to elucidate much.